ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/AutoMediaBrowserService.kt

```diff
@@ -0,0 +1,186 @@
+package org.moire.ultrasonic.service
+
+import android.os.Bundle
+import android.os.Handler
+import android.support.v4.media.MediaBrowserCompat
+import android.support.v4.media.MediaDescriptionCompat
+import android.support.v4.media.session.MediaSessionCompat
+import androidx.media.MediaBrowserServiceCompat
+import androidx.media.utils.MediaConstants
+import org.koin.android.ext.android.inject
+import org.moire.ultrasonic.util.MediaSessionEventDistributor
+import org.moire.ultrasonic.util.MediaSessionEventListener
+import timber.log.Timber
+
+
+const val MY_MEDIA_ROOT_ID = "MY_MEDIA_ROOT_ID"
+const val MY_MEDIA_ALBUM_ID = "MY_MEDIA_ALBUM_ID"
+const val MY_MEDIA_ARTIST_ID = "MY_MEDIA_ARTIST_ID"
+const val MY_MEDIA_ALBUM_ITEM = "MY_MEDIA_ALBUM_ITEM"
+const val MY_MEDIA_LIBRARY_ID = "MY_MEDIA_LIBRARY_ID"
+const val MY_MEDIA_PLAYLIST_ID = "MY_MEDIA_PLAYLIST_ID"
+
+class AutoMediaBrowserService : MediaBrowserServiceCompat() {
+
+    private lateinit var mediaSessionEventListener: MediaSessionEventListener
+    private val mediaSessionEventDistributor: MediaSessionEventDistributor by inject()
+    private val lifecycleSupport: MediaPlayerLifecycleSupport by inject()
+
+    override fun onCreate() {
+        super.onCreate()
+
+        mediaSessionEventListener = object : MediaSessionEventListener {
+            override fun onMediaSessionTokenCreated(token: MediaSessionCompat.Token) {
+                Timber.i("AutoMediaBrowserService onMediaSessionTokenCreated called")
+                if (sessionToken == null) {
+                    Timber.i("AutoMediaBrowserService onMediaSessionTokenCreated session token was null, set it to %s", token.toString())
+                    sessionToken = token
+                }
+            }
+
+            override fun onPlayFromMediaIdRequested(mediaId: String?, extras: Bundle?) {
+                // TODO implement
+                Timber.i("AutoMediaBrowserService onPlayFromMediaIdRequested called")
+            }
+
+            override fun onPlayFromSearchRequested(query: String?, extras: Bundle?) {
+                // TODO implement
+                Timber.i("AutoMediaBrowserService onPlayFromSearchRequested called")
+            }
+        }
+
+        mediaSessionEventDistributor.subscribe(mediaSessionEventListener)
+
+        val handler = Handler()
+        handler.postDelayed({
```

Comment:
THIS! This right here is gold! You are a super hero @nitehu ! I tried this out and it solved my problem of the out-of-order `onCreate`. I'm afraid I didn't have much more time to look at everything else you did, I'll keep looking. You've done a bunch that I don't understand, but I think I've handled some of the parts that are just stubs in your code (the browsing parts). Maybe we can join them up. There might be a solution lurking in here :)

I'm optimistic again. Did you manage to get a working dev environment? You must have to get all this code done. Is it using hardware or did you get it done with an emulator? It's a pain to set up isn't it?

Reply:
Yes I've set up the dev environment based on the link you shared. I searched all around the web but it seems this won't work with an emulator, just with a real device. Of course I hate this solution already, doing all the magic rituals of debug service starting, adb port forward, etc. everytime I want to start testing... XD

I'm currently working on the tricky Now Playing parts, I also think that the browsing is more straightforward and can use the solutions you've already implemented.